remote-server SSH wrapper install fails on macOS due to bash-3.2 tilde expansion

### Pre-submit Checks

- [x] I have searched Warp bugs and there are no duplicates
- [x] I have searched Warp known issues page and my issue is not there
- [x] I have included the logs

### Describe the bug

Warp's remote-server SSH wrapper (the new agent-based one, not the legacy `warpify`/tmux wrapper) fails to install its `oz` agent binary on any remote macOS host where `bash` resolves to Apple's stock `/bin/bash` (version 3.2.57 — i.e. every macOS host that hasn't put homebrew bash on PATH for non-login shells, which is the default).

User-visible symptom in `~/Library/Logs/warp.log`:

```
[INFO]  Checking for remote server binary at ~/.warp/remote-server/oz-v0.2026.04.29.08.57.stable_01
[INFO]  Installing remote server binary to ~/.warp/remote-server/oz-v0.2026.04.29.08.57.stable_01
[INFO]  Starting remote server connection for session SessionId(...)
[INFO]  [remote_server] zsh:1: no such file or directory: /Users/<user>/.warp/remote-server/oz-v0.2026.04.29.08.57.stable_01
[INFO]  Reader task: server disconnected (EOF)
[ERROR] Connection failed for session SessionId(...): initialize: Response channel closed before receiving a reply
[WARN]  Remote server connection failed for session SessionId(...); flushing bootstrap to unblock SSH session
```

Warp falls back to ControlMaster/legacy mode and the user sees `Initialize: Response channel closed before receiving a reply` — a misleading symptom (the protocol channel didn't close prematurely; the agent simply doesn't exist).

### Root cause

`crates/remote_server/src/install_remote_server.sh:30`:

```bash
install_dir="${install_dir/#\~/"$HOME"}"
```

`run_ssh_script` in `crates/remote_server/src/ssh.rs` invokes the install script with `ssh ... bash -s` and pipes the script body into stdin. On macOS the resolved `bash` is `/bin/bash` (3.2.57 — Apple has not shipped a newer bash since the GPLv3 transition).

In bash 3.2, the `"$HOME"` *inside the replacement* of `${var/pattern/replacement}` is treated as **6 literal characters including the double-quotes**, not expanded. Bash 4+ strips them. Empirical confirmation:

```
$ /bin/bash --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin25)

$ HOME=/Users/me /bin/bash -c 'x=~/foo; echo "${x/#\~/"$HOME"}"'
"/Users/me"/foo                      # ← literal quotes

$ HOME=/Users/me /bin/bash -c 'x=~/foo; echo "${x/#\~/$HOME}"'
/Users/me/foo                        # ← correct
```

Downstream consequences in the install script:

- `mkdir -p '"/Users/<user>"/.warp/remote-server'` succeeds, creating a directory tree under `$HOME` whose first component is literally `"`.
- The 278 MB agent tarball downloads and extracts cleanly.
- `mv` lands the binary at `$HOME/"/Users/<user>"/.warp/remote-server/oz-v0...`.
- `install_binary()` returns `Ok(())`.
- `connect()` then asks zsh to exec `~/.warp/remote-server/oz-v0...` (the *correct* path), which doesn't exist → `zsh:1: no such file or directory`.

You can verify on any affected host: there will be a directory literally named `"` in the user's home folder containing a misplaced `oz-v0...` binary. On my Mac Studio that directory was dated to my very first connection attempt — the bug has been silent since the first install.

### To reproduce

1. From a Mac running Warp, `ssh` into another Mac (or any host where `bash -s` resolves to bash 3.2 — i.e. any macOS without `/opt/homebrew/bin/bash` shadowed via login shell config).
2. Wait for the remote-server bootstrap.
3. Observe the `initialize: Response channel closed before receiving a reply` error and the misplaced binary at `~/"/Users/<user>"/.warp/remote-server/oz-v0...` on the remote.

### Expected behavior

Install lands at `$HOME/.warp/remote-server/oz-v0...`; remote-server agent launches; full Warp UI works over SSH.

### Proposed fix

One character: drop the inner quotes around `$HOME` in the replacement. Bash 3.2 and 4+ both expand `${install_dir/#\~/$HOME}` correctly.

Happy to open the PR — I have it ready locally with two regression tests:

1. A live test that pipes the relevant snippet into `/bin/bash` and asserts the resolved path contains no literal `"` characters.
2. A static check over `INSTALL_SCRIPT_TEMPLATE` that grep-matches the buggy `/#\~/"$HOME"` pattern and explains why in the failure message.

### Operating system (OS)

macOS

### Operating system and version

macOS 26.2 (build 25C56), Darwin 25.2.0 arm64 (Mac Studio M2 Max). Reproduces against any Mac running default `/bin/bash` 3.2.57.

### Shell Version

Local: zsh 5.9. Remote `bash --version`: `GNU bash, version 3.2.57(1)-release (arm64-apple-darwin25)`.

### Current Warp version

v0.2026.04.29.08.57.stable_01

### Regression

No, this bug or issue has existed throughout my experience using Warp

### Additional context

- The connection works fine when Warp falls back to the legacy ControlMaster path. The bug only affects the new remote-server wrapper.
- Workaround for affected users: pre-place the agent at the correct path so `binary_check_command` (`test -x ~/.warp/remote-server/oz-v0...`) returns 0 and the broken install step is skipped:
  ```
  ssh <host> 'mv "\"/$HOME\"/.warp/remote-server/oz-v0..." ~/.warp/remote-server/oz-v0...'
  ```
- The bug doesn't happen on Linux remotes because Linux bash is 4+ (Apple is the outlier — they freeze /bin/bash at 3.2 to avoid GPLv3).
- Diagnostic timeline I went through, in case useful for triage: shell-init noise from missing brew tools (red herring — `bash -s` doesn't source rc files); URL guessing against `releases.warp.dev` (wrong base — actual URL is `app.warp.dev/download/cli`, served from GCS); `fs_usage`-style polling watcher confirmed zero filesystem writes in `~/.warp/remote-server/` during failure (red herring — writes were going to `~/"/Users/.../"...`); reading `crates/remote_server/src/install_remote_server.sh` and `ssh.rs:run_ssh_script` made the bash-3.2 incompatibility obvious.

### Does this block you from using Warp daily?

No

### Is this an issue only in Warp?

Yes, I confirmed that this only happens in Warp, not other terminals.

--- a/crates/remote_server/src/install_remote_server.sh
+++ b/crates/remote_server/src/install_remote_server.sh
@@ -27,12 +27,20 @@ case "$os_kernel" in
 esac
 
 install_dir="{install_dir}"
-# Inner quotes around `$HOME` would be safe on bash 4+, but on macOS's
-# stock /bin/bash 3.2 they are taken as literal characters in the
-# replacement, producing a path like `"/Users/<user>"/.warp/...` and
-# silently steering the install into a directory tree literally named
-# `"`. Pipe the script through bash 3.2 before changing this line.
-install_dir="${install_dir/#\~/$HOME}"
+# Avoid `${var/pattern/replacement}` for tilde expansion. Two
+# interpreter quirks make it dangerous in this script:
+#   1. bash 3.2 (macOS /bin/bash) keeps inner double-quotes around the
+#      replacement literal, so `"$HOME"` ends up as 6 literal
+#      characters and the install lands under a directory tree
+#      literally named `"`.
+#   2. bash 5.2+ enables `patsub_replacement` by default, which makes
+#      `&` in the replacement expand to the matched pattern, so a
+#      `$HOME` containing `&` resolves to a `~`-substituted path.
+# Use `case` + `${var#\~}` instead — works on bash 3.2 and bash 5.2+
+# without surprises.
+case "$install_dir" in
+  "~"|"~/"*) install_dir="${HOME}${install_dir#\~}" ;;
+esac
 mkdir -p "$install_dir"
 
 tmpdir=$(mktemp -d "$install_dir/.install.XXXXXX")
@@ -50,8 +58,10 @@ trap cleanup EXIT
 staging_tarball_path="{staging_tarball_path}"
 if [ -n "$staging_tarball_path" ]; then
   # SCP fallback: tarball already uploaded by the client.
-  # Same bash 3.2 caveat as the install_dir tilde expansion above.
-  staging_tarball_path="${staging_tarball_path/#\~/$HOME}"
+  # Same tilde-expansion caveat as install_dir above.
+  case "$staging_tarball_path" in
+    "~"|"~/"*) staging_tarball_path="${HOME}${staging_tarball_path#\~}" ;;
+  esac
   mv "$staging_tarball_path" "$tmpdir/oz.tar.gz"
 else
   # Normal path: download via curl or wget.

--- a/crates/remote_server/src/install_remote_server.sh
+++ b/crates/remote_server/src/install_remote_server.sh
@@ -27,7 +27,12 @@ case "$os_kernel" in
 esac
 
 install_dir="{install_dir}"
-install_dir="${install_dir/#\~/"$HOME"}"
+# Inner quotes around `$HOME` would be safe on bash 4+, but on macOS's
+# stock /bin/bash 3.2 they are taken as literal characters in the
+# replacement, producing a path like `"/Users/<user>"/.warp/...` and
+# silently steering the install into a directory tree literally named
+# `"`. Pipe the script through bash 3.2 before changing this line.
+install_dir="${install_dir/#\~/$HOME}"
 mkdir -p "$install_dir"
 
 tmpdir=$(mktemp -d "$install_dir/.install.XXXXXX")

--- a/crates/remote_server/src/install_remote_server.sh
+++ b/crates/remote_server/src/install_remote_server.sh
@@ -50,7 +50,8 @@ trap cleanup EXIT
 staging_tarball_path="{staging_tarball_path}"
 if [ -n "$staging_tarball_path" ]; then
   # SCP fallback: tarball already uploaded by the client.
-  staging_tarball_path="${staging_tarball_path/#\~/"$HOME"}"
+  # Same bash 3.2 caveat as the install_dir tilde expansion above.
+  staging_tarball_path="${staging_tarball_path/#\~/$HOME}"
   mv "$staging_tarball_path" "$tmpdir/oz.tar.gz"
 else
   # Normal path: download via curl or wget.

--- a/crates/remote_server/src/setup_tests.rs
+++ b/crates/remote_server/src/setup_tests.rs
@@ -170,26 +170,31 @@ fn parse_preinstall_unsupported_non_glibc() {
     assert!(!result.is_supported());
 }
 
-/// Regression: the install script's tilde-expansion line must work in
-/// macOS's stock `/bin/bash` 3.2.57. In bash 3.2, `"$HOME"` inside the
-/// replacement of `${var/pattern/replacement}` is treated as 6 literal
-/// characters (the quotes are not stripped), which turned
-/// `~/.warp/remote-server` into the *relative* path
-/// `"/Users/<user>"/.warp/remote-server` and silently steered the
-/// install into a directory tree literally named `"`. The launch step
-/// then looked at the real `$HOME/.warp/remote-server/oz-...`, found
-/// nothing, and reported "no such file or directory" → "Response
-/// channel closed".
+/// Regression: the install script's tilde-expansion logic must work
+/// across the bash versions we actually invoke at install time
+/// (`run_ssh_script` pipes the script into `bash -s` on the remote).
+/// Two interpreter quirks have to be avoided simultaneously:
+///
+///   1. bash 3.2 (macOS `/bin/bash`) keeps inner double-quotes around
+///      the replacement of `${var/pattern/replacement}` literal, so
+///      `"$HOME"` ends up as 6 literal characters and the install
+///      lands under a directory tree literally named `"`.
+///   2. bash 5.2+ with `patsub_replacement` (default-on) treats `&`
+///      in the replacement as the matched pattern, so a `$HOME`
+///      containing `&` resolves to a `~`-substituted path.
+///
+/// Both bugs surface as the install binary landing somewhere Warp's
+/// launch step doesn't look, producing a misleading "Response channel
+/// closed before receiving a reply".
 ///
 /// This test drives the *actual* production script (via
-/// [`install_script`]) rather than a hand-copied snippet so a future
-/// change that reintroduces the quoted form can't slip past by being
-/// dissimilar enough to dodge the static-grep check below. We
-/// truncate just before `mkdir -p` so no filesystem side effects leak
-/// out of the test, and append a marker `printf` to capture the
-/// resolved `install_dir`.
+/// [`install_script`]) rather than a hand-copied snippet, and runs it
+/// against several `HOME` values to exercise the patsub-`&` trap as
+/// well as the quote-literal trap. We truncate just before `mkdir -p`
+/// so no filesystem side effects leak out of the test, and append a
+/// marker `printf` to capture the resolved `install_dir`.
 #[test]
-fn install_script_tilde_expansion_works_in_bash_3_2() {
+fn install_script_tilde_expansion_resolves_correctly() {
     use std::process::{Command, Stdio};
 
     let bash = if std::path::Path::new("/bin/bash").exists() {
@@ -209,78 +214,79 @@ fn install_script_tilde_expansion_works_in_bash_3_2() {
         prefix = &script[..cutoff],
     );
 
-    let fake_home = "/Users/test";
-    let output = Command::new(bash)
-        .arg("-c")
-        .arg(&probe)
-        .env("HOME", fake_home)
-        .stdout(Stdio::piped())
-        .stderr(Stdio::piped())
-        .output()
-        .expect("failed to spawn bash");
+    // Run the probe against a matrix of HOME values. The first is an
+    // ordinary path; the second contains `&`, which exercises bash
+    // 5.2's patsub_replacement (where it would otherwise expand to
+    // the matched `~`).
+    let cases = [
+        ("/Users/test", "ordinary HOME"),
+        (
+            "/Users/A&B",
+            "HOME with `&` (bash 5.2 patsub_replacement trap)",
+        ),
+    ];
 
-    assert!(
-        output.status.success(),
-        "bash exited with {:?}: stderr={}",
-        output.status,
-        String::from_utf8_lossy(&output.stderr),
-    );
+    for (fake_home, label) in cases {
+        let output = Command::new(bash)
+            .arg("-c")
+            .arg(&probe)
+            .env("HOME", fake_home)
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .output()
+            .expect("failed to spawn bash");
 
-    let install_dir = String::from_utf8_lossy(&output.stdout);
-    assert!(
-        !install_dir.contains('"'),
-        "install_dir contains literal quote characters \
-         (bash 3.2 quote-literal regression): {install_dir:?}",
-    );
+        assert!(
+            output.status.success(),
+            "[{label}] bash exited with {:?}: stderr={}",
+            output.status,
+            String::from_utf8_lossy(&output.stderr),
+        );
 
-    // Cross-check against the production layout: tilde must resolve
-    // to HOME, so the result equals `remote_server_dir()` with the
-    // leading `~` replaced.
-    let expected = remote_server_dir().replacen('~', fake_home, 1);
-    assert_eq!(install_dir, expected);
+        let install_dir = String::from_utf8_lossy(&output.stdout);
+        assert!(
+            !install_dir.contains('"'),
+            "[{label}] install_dir contains literal quote characters \
+             (bash 3.2 quote-literal regression): {install_dir:?}",
+        );
+
+        // Cross-check against the production layout: tilde must
+        // resolve to HOME, so the result equals `remote_server_dir()`
+        // with the leading `~` replaced.
+        let expected = remote_server_dir().replacen('~', fake_home, 1);
+        assert_eq!(
+            install_dir, expected,
+            "[{label}] install_dir resolved incorrectly",
+        );
+    }
 }
 
-/// Regression: guards against re-introducing any quoted form of the
-/// tilde substitution by scanning the script source itself.
-/// Complements [`install_script_tilde_expansion_works_in_bash_3_2`] —
-/// the live bash test catches behavioural regressions, this static
-/// check catches them earlier and rules out variants the live test
-/// might miss (`"${HOME}"`, `'$HOME'`, etc.).
-///
-/// The rule: every `${var/#\~/...}` occurrence in the script must
-/// have its replacement start with `$` (i.e. `$HOME` or `${HOME}`
-/// directly). Anything else is a quoted form that bash 3.2 will
-/// preserve as literal characters.
+/// Regression: guards against re-introducing the
+/// `${var/pattern/replacement}` tilde-substitution form, which has two
+/// known interpreter bugs (see
+/// [`install_script_tilde_expansion_resolves_correctly`] for details).
+/// Complements the live bash test — the live test catches behavioural
+/// regressions, this static check fails fast and explains *why* in
+/// the assertion message so a future contributor doesn't have to
+/// re-discover the constraints from a CI failure.
 #[test]
-fn install_script_does_not_quote_home_in_tilde_substitution() {
+fn install_script_avoids_pattern_substitution_for_tilde_expansion() {
     let template = INSTALL_SCRIPT_TEMPLATE;
-    let needle = r"/#\~/";
-    let mut hits = 0;
-    let mut search_from = 0;
-    while let Some(off) = template[search_from..].find(needle) {
-        let abs = search_from + off;
-        let after = &template[abs + needle.len()..];
-        let next = after.chars().next();
-        let context_end = abs + needle.len() + after.len().min(16);
-        let context = &template[abs..context_end];
-        assert_eq!(
-            next,
-            Some('$'),
-            "install_remote_server.sh has tilde substitution `{context}` \
-             at byte {abs}; the replacement must start with `$` (i.e. \
-             `$HOME` or `${{HOME}}`) — not a quote, single quote, or \
-             anything else — because bash 3.2 (macOS /bin/bash) keeps \
-             the surrounding characters literal in the replacement of \
-             `${{var/pattern/replacement}}`",
-        );
-        hits += 1;
-        search_from = abs + needle.len();
-    }
     assert!(
-        hits >= 1,
-        "no `/#\\~/...` substitutions found in install_remote_server.sh \
-         — has the script changed shape such that this test is no \
-         longer guarding what it claims to?",
+        !template.contains(r"/#\~/"),
+        "install_remote_server.sh uses `${{var/#\\~/...}}` for tilde \
+         expansion. This form has two known interpreter bugs that \
+         silently mis-resolve the install path:\n\
+         \n\
+           1. bash 3.2 (macOS /bin/bash) keeps inner double-quotes \
+              around the replacement literal, so `\"$HOME\"` ends up \
+              as 6 literal characters including the quotes.\n\
+           2. bash 5.2+ enables `patsub_replacement` by default, which \
+              makes `&` in the replacement expand to the matched \
+              pattern, so a `$HOME` containing `&` resolves wrong.\n\
+         \n\
+         Use `case`/`${{var#\\~}}` instead — see install_remote_server.sh \
+         for the pattern.",
     );
 }
 

--- a/crates/remote_server/src/setup_tests.rs
+++ b/crates/remote_server/src/setup_tests.rs
@@ -195,7 +195,8 @@ fn parse_preinstall_unsupported_non_glibc() {
 /// marker `printf` to capture the resolved `install_dir`.
 #[test]
 fn install_script_tilde_expansion_resolves_correctly() {
-    use std::process::{Command, Stdio};
+    use command::blocking::Command;
+    use std::process::Stdio;
 
     let bash = if std::path::Path::new("/bin/bash").exists() {
         "/bin/bash"

--- a/crates/remote_server/src/setup_tests.rs
+++ b/crates/remote_server/src/setup_tests.rs
@@ -170,6 +170,88 @@ fn parse_preinstall_unsupported_non_glibc() {
     assert!(!result.is_supported());
 }
 
+/// Regression: the install script's tilde-expansion line must work in
+/// macOS's stock `/bin/bash` 3.2.57. In bash 3.2, `"$HOME"` inside the
+/// replacement of `${var/pattern/replacement}` is treated as 6 literal
+/// characters (the quotes are not stripped), which turned
+/// `~/.warp/remote-server` into the *relative* path
+/// `"/Users/<user>"/.warp/remote-server` and silently steered the
+/// install into a directory tree literally named `"`. The launch step
+/// then looked at the real `$HOME/.warp/remote-server/oz-...`, found
+/// nothing, and reported "no such file or directory" → "Response
+/// channel closed".
+///
+/// Running the materialised script under `/bin/bash` ensures the
+/// expansion is correct on the bash version we actually invoke at
+/// install time (`run_ssh_script` pipes into `bash -s`, which on macOS
+/// is bash 3.2).
+#[test]
+fn install_script_tilde_expansion_works_in_bash_3_2() {
+    use std::process::{Command, Stdio};
+
+    let bash = if std::path::Path::new("/bin/bash").exists() {
+        "/bin/bash"
+    } else {
+        "bash"
+    };
+
+    // Inline only the lines from install_remote_server.sh that resolve
+    // the install directory, then echo the result. Keeps the test
+    // independent of the network-bound steps further down the script.
+    let snippet = r#"
+        set -e
+        install_dir="~/.warp/remote-server"
+        install_dir="${install_dir/#\~/$HOME}"
+        printf '%s' "$install_dir"
+    "#;
+
+    let output = Command::new(bash)
+        .arg("-c")
+        .arg(snippet)
+        .env("HOME", "/Users/test")
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .output()
+        .expect("failed to spawn bash");
+
+    assert!(
+        output.status.success(),
+        "bash exited with {:?}: stderr={}",
+        output.status,
+        String::from_utf8_lossy(&output.stderr),
+    );
+
+    let install_dir = String::from_utf8_lossy(&output.stdout);
+    assert_eq!(
+        install_dir, "/Users/test/.warp/remote-server",
+        "tilde expansion produced wrong path; \
+         likely a regression of the bash-3.2 quote-literal bug",
+    );
+    assert!(
+        !install_dir.contains('"'),
+        "install_dir contains literal quote characters \
+         (bash 3.2 quote-literal regression): {install_dir:?}",
+    );
+}
+
+/// Regression: guards against re-introducing the literal-quotes form
+/// of the tilde substitution by scanning the script source itself.
+/// Complements `install_script_tilde_expansion_works_in_bash_3_2` —
+/// the live bash test catches behavioural regressions, this static
+/// check catches them earlier and explains *why* in the failure
+/// message.
+#[test]
+fn install_script_does_not_quote_home_in_tilde_substitution() {
+    let template = INSTALL_SCRIPT_TEMPLATE;
+    assert!(
+        !template.contains("/#\\~/\"$HOME\""),
+        "install_remote_server.sh uses `${{var/#\\~/\"$HOME\"}}`, \
+         which on bash 3.2 (macOS /bin/bash) substitutes the literal \
+         characters `\"$HOME\"` instead of the expanded value. Use \
+         `${{var/#\\~/$HOME}}` (no inner quotes) instead.",
+    );
+}
+
 #[test]
 fn parse_preinstall_missing_status_falls_open() {
     // Garbled / partial script output — missing status field. Confirms

--- a/crates/remote_server/src/setup_tests.rs
+++ b/crates/remote_server/src/setup_tests.rs
@@ -181,10 +181,13 @@ fn parse_preinstall_unsupported_non_glibc() {
 /// nothing, and reported "no such file or directory" → "Response
 /// channel closed".
 ///
-/// Running the materialised script under `/bin/bash` ensures the
-/// expansion is correct on the bash version we actually invoke at
-/// install time (`run_ssh_script` pipes into `bash -s`, which on macOS
-/// is bash 3.2).
+/// This test drives the *actual* production script (via
+/// [`install_script`]) rather than a hand-copied snippet so a future
+/// change that reintroduces the quoted form can't slip past by being
+/// dissimilar enough to dodge the static-grep check below. We
+/// truncate just before `mkdir -p` so no filesystem side effects leak
+/// out of the test, and append a marker `printf` to capture the
+/// resolved `install_dir`.
 #[test]
 fn install_script_tilde_expansion_works_in_bash_3_2() {
     use std::process::{Command, Stdio};
@@ -195,20 +198,22 @@ fn install_script_tilde_expansion_works_in_bash_3_2() {
         "bash"
     };
 
-    // Inline only the lines from install_remote_server.sh that resolve
-    // the install directory, then echo the result. Keeps the test
-    // independent of the network-bound steps further down the script.
-    let snippet = r#"
-        set -e
-        install_dir="~/.warp/remote-server"
-        install_dir="${install_dir/#\~/$HOME}"
-        printf '%s' "$install_dir"
-    "#;
+    let script = install_script(None);
+    let cutoff = script.find("mkdir -p \"$install_dir\"").expect(
+        "install script no longer contains the `mkdir -p \"$install_dir\"` \
+         checkpoint this test relies on; update the test alongside the \
+         script change",
+    );
+    let probe = format!(
+        "{prefix}\nprintf '%s' \"$install_dir\"\nexit 0\n",
+        prefix = &script[..cutoff],
+    );
 
+    let fake_home = "/Users/test";
     let output = Command::new(bash)
         .arg("-c")
-        .arg(snippet)
-        .env("HOME", "/Users/test")
+        .arg(&probe)
+        .env("HOME", fake_home)
         .stdout(Stdio::piped())
         .stderr(Stdio::piped())
         .output()
@@ -222,33 +227,60 @@ fn install_script_tilde_expansion_works_in_bash_3_2() {
     );
 
     let install_dir = String::from_utf8_lossy(&output.stdout);
-    assert_eq!(
-        install_dir, "/Users/test/.warp/remote-server",
-        "tilde expansion produced wrong path; \
-         likely a regression of the bash-3.2 quote-literal bug",
-    );
     assert!(
         !install_dir.contains('"'),
         "install_dir contains literal quote characters \
          (bash 3.2 quote-literal regression): {install_dir:?}",
     );
+
+    // Cross-check against the production layout: tilde must resolve
+    // to HOME, so the result equals `remote_server_dir()` with the
+    // leading `~` replaced.
+    let expected = remote_server_dir().replacen('~', fake_home, 1);
+    assert_eq!(install_dir, expected);
 }
 
-/// Regression: guards against re-introducing the literal-quotes form
-/// of the tilde substitution by scanning the script source itself.
-/// Complements `install_script_tilde_expansion_works_in_bash_3_2` —
+/// Regression: guards against re-introducing any quoted form of the
+/// tilde substitution by scanning the script source itself.
+/// Complements [`install_script_tilde_expansion_works_in_bash_3_2`] —
 /// the live bash test catches behavioural regressions, this static
-/// check catches them earlier and explains *why* in the failure
-/// message.
+/// check catches them earlier and rules out variants the live test
+/// might miss (`"${HOME}"`, `'$HOME'`, etc.).
+///
+/// The rule: every `${var/#\~/...}` occurrence in the script must
+/// have its replacement start with `$` (i.e. `$HOME` or `${HOME}`
+/// directly). Anything else is a quoted form that bash 3.2 will
+/// preserve as literal characters.
 #[test]
 fn install_script_does_not_quote_home_in_tilde_substitution() {
     let template = INSTALL_SCRIPT_TEMPLATE;
+    let needle = r"/#\~/";
+    let mut hits = 0;
+    let mut search_from = 0;
+    while let Some(off) = template[search_from..].find(needle) {
+        let abs = search_from + off;
+        let after = &template[abs + needle.len()..];
+        let next = after.chars().next();
+        let context_end = abs + needle.len() + after.len().min(16);
+        let context = &template[abs..context_end];
+        assert_eq!(
+            next,
+            Some('$'),
+            "install_remote_server.sh has tilde substitution `{context}` \
+             at byte {abs}; the replacement must start with `$` (i.e. \
+             `$HOME` or `${{HOME}}`) — not a quote, single quote, or \
+             anything else — because bash 3.2 (macOS /bin/bash) keeps \
+             the surrounding characters literal in the replacement of \
+             `${{var/pattern/replacement}}`",
+        );
+        hits += 1;
+        search_from = abs + needle.len();
+    }
     assert!(
-        !template.contains("/#\\~/\"$HOME\""),
-        "install_remote_server.sh uses `${{var/#\\~/\"$HOME\"}}`, \
-         which on bash 3.2 (macOS /bin/bash) substitutes the literal \
-         characters `\"$HOME\"` instead of the expanded value. Use \
-         `${{var/#\\~/$HOME}}` (no inner quotes) instead.",
+        hits >= 1,
+        "no `/#\\~/...` substitutions found in install_remote_server.sh \
+         — has the script changed shape such that this test is no \
+         longer guarding what it claims to?",
     );
 }
 

--- a/crates/remote_server/src/setup_tests.rs
+++ b/crates/remote_server/src/setup_tests.rs
@@ -193,6 +193,11 @@ fn parse_preinstall_unsupported_non_glibc() {
 /// well as the quote-literal trap. We truncate just before `mkdir -p`
 /// so no filesystem side effects leak out of the test, and append a
 /// marker `printf` to capture the resolved `install_dir`.
+///
+/// Gated to Unix because the test invokes `/bin/bash` (or `bash` from
+/// PATH) directly. The bug only matters on Unix remotes anyway —
+/// Warp's remote-server SSH wrapper doesn't target Windows hosts.
+#[cfg(unix)]
 #[test]
 fn install_script_tilde_expansion_resolves_correctly() {
     use command::blocking::Command;

